mkosi/vmspawn.py

```diff
@@ -0,0 +1,227 @@
+    start_swtpm,
+    start_virtiofsd,
+    vsock_notify_handler,
+)
+from mkosi.run import run, spawn
+from mkosi.types import PathString
+from mkosi.util import INVOKING_USER
+
+
+def config_feature_to_tristate(cf: ConfigFeature) -> str:
+    if cf == ConfigFeature.enabled:
+        return "yes"
+    if cf == ConfigFeature.disabled:
+        return "no"
+    return ""
+
+
+def run_vmspawn(args: MkosiArgs, config: MkosiConfig, qemu_device_fds: Mapping[QemuDeviceNode, int]) -> None:
+    if config.output_format != OutputFormat.disk:
+        die(f"{config.output_format} images cannot be booted in systemd-vmspawn")
+
+    if (config.runtime_trees and config.qemu_firmware == QemuFirmware.bios):
+        die("RuntimeTrees= cannot be used when booting in BIOS firmware")
+
+    if config.qemu_kvm == ConfigFeature.enabled and not QemuDeviceNode.kvm.available():
+        die("KVM acceleration requested but cannot access /dev/kvm")
+
+    if config.qemu_vsock == ConfigFeature.enabled and QemuDeviceNode.vhost_vsock not in qemu_device_fds:
+        die("VSock requested but cannot access /dev/vhost-vsock")
+
+    if config.qemu_cdrom:
+        die("systemd-vmspawn does not support CD-ROM images")
+
+    if config.qemu_kernel:
+        kernel = config.qemu_kernel
+    elif "-kernel" in args.cmdline:
+        kernel = Path(args.cmdline[args.cmdline.index("-kernel") + 1])
+    else:
+        kernel = None
+
+    if kernel and not kernel.exists():
+        die(f"Kernel not found at {kernel}")
+
+    if config.qemu_firmware == QemuFirmware.auto:
+        if kernel:
+            firmware = QemuFirmware.uefi if KernelType.identify(kernel) != KernelType.unknown else QemuFirmware.linux
+        elif (
+            config.output_format in (OutputFormat.cpio, OutputFormat.directory) or
+            config.architecture.to_efi() is None
+        ):
+            firmware = QemuFirmware.linux
+        else:
+            firmware = QemuFirmware.uefi
+    else:
+        firmware = config.qemu_firmware
+
+    if firmware == QemuFirmware.linux:
+        kernel = config.output_dir_or_cwd() / config.output_split_kernel
+        if not kernel.exists():
+            die(
+                f"Kernel or UKI not found at {kernel}, please install a kernel in the image "
+                "or provide a -kernel argument to mkosi vmspawn"
+            )
+
+    notifications: dict[str, str] = {}
+    cmdline: list[PathString] = [
+        "systemd-vmspawn",
+        f"--qemu-smp={config.qemu_smp}",
+        f"--qemu-mem={config.qemu_mem}",
+        f"--qemu-kvm={config_feature_to_tristate(config.qemu_kvm)}",
+
+        # we need to handle this ourselves as we cannot pass the fd to vmspawn
+        # directly but have to pass it to qemu instead
+        # to prevent a conflict we disable vsock for vmspawn
+        "--qemu-vsock=no",
+    ]
+
+    if config.qemu_gui:
+        cmdline.append("--qemu-gui")
+
+    if config.secure_boot:
+        cmdline.append("--secure-boot=yes")
```

Comment:
The secure boot option is only for building images and shouldn't be used when booting them

Reply:
For vmspawn this affects which kind of firmware it will find.
Can images built for secure boot be booted with non secure-boot firmware?
If not I think it is best to leave this in so that vmspawn will pick a secure-boot firmware if one exists.